Battery widget

marriot2018 opened the pull request:
Here is the creation of the battery widget using qml and c++

marriot2018 committed:
project creation

marriot2018 committed:
batteryqml

marriot2018 committed:
battery.h

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+import QtQuick.Controls
+import MyBattery 1.0
+
+Window {
```
you are creating a separate window, let's change this and make an instance of BatteryWidget in main instead, please

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+
+    function progressBarFill(event) {
+        if (event.key === Qt.Key_Q) {
+            batteryInfo.increased(fillRectangle.height)
```
although this logic works, wouldn't it make more sense that increase function works with percentages instead of a UI element height? And to calculate the height of UI element based on backend percentage?

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+
+    Battery {
+        id: batteryInfo
+        onIncrease: {
```
if you change the logic to use percentage instead, you will only need one handler "onPercentageChanged"

Marisol122345 replied:
Or even easier, you can have 2 Q_PROPERTY. One with percentage and another one with color. On UI side you would only need to have height calculation bound to percentage Q_PROPERTY and color bound to color Q_PROPERTY

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+        id: batteryAnimationUp
+        target: batteryWidget
+        property: "batteryValue"
+        duration: 250
```
for repeated values it is probably better to set a property that you can reuse

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+        duration: 250
+    }
+
+    PropertyAnimation {
```
do we need 2 animations that do the same thing?

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+        width: rectangleBorder.width
+        height: batteryValue
+
+        Keys.onPressed: (event)=> {
```
this wouldn't be necessary with percentage approach

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    Rectangle {
+        anchors.bottom: rectangleBorder.top
```
when having multiple anchors, it is better to group them

```suggestion
        anchors {
               bottom: rectangleBorder.top
               horizontalCenter: rectangleBorder.horizontalCenter
        }
```

Marisol122345 commented on battery/BatteryWidget.qml:
```diff
@@ -0,0 +1,84 @@
+        height: 5
+    }
+
+    Component.onCompleted: {
```
also not needed with percentage approach

Marisol122345 commented on battery/battery.cpp:
```diff
@@ -0,0 +1,73 @@
+
+Battery::Battery(QObject *parent) : QObject{parent}
+{
+    m_value = 0;
```
you can initiliaze in initializer list

Marisol122345 commented on battery/main.cpp:
```diff
@@ -0,0 +1,21 @@
+    QGuiApplication app(argc, argv);
+
+    QQmlApplicationEngine engine;
+    qmlRegisterType<Battery>("MyBattery", 1, 0, "Battery");
```
this option is good. In the case you needed a singleton you could also use setContextProperty

Marisol122345 commented on battery/battery.h:
```diff
@@ -0,0 +1,44 @@
+class Battery : public QObject
+{
+    Q_OBJECT
+    Q_PROPERTY(int batteryCharge READ getValue WRITE setValue)
```
if you add NOTIFY to your Q_PROPERTY, the property (in this case batteryCharge can be "observed" by QML and it will be notified whenever a change is made

Marisol122345 replied:
I also suggest to use more descriptive words, setValue could be any value

```suggestion
    Q_PROPERTY(int batteryPercentage READ getBatteryPercentage WRITE setBatteryPercentage NOTIFY batteryPercentageChanged)
```

marriot2018 committed:
Fix

I have updated the name of the QProperty to BatteryPercentage, also the respective getter ansd setter.
Now I only use one propertyAnimation
BatteryWidget is now instantiated in the Main.qml

marriot2018 committed:
Height value fix

In this commit now the height for the percentage bar will come from the backend

marriot2018 committed:
decimals fix

marriot2018 committed:
fix round decimals

marriot2018 committed:
setBatteryPercentage fix

the decrease and increase functions were deleted, and now the setBatteryPercentage does te addition and substraction

marriot2018 committed:
Initializer List

marriot2018 committed:
Bar Color

marriot2018 committed:
batteryAnimation
